src/persist-client/src/internal/compact.rs

```diff
@@ -254,14 +275,233 @@ impl Compactor {
+        let mut update_buffer_size_bytes = 0;
+        let mut greatest_kv: Option<(Vec<u8>, Vec<u8>)> = None;
+
+        // we'll reference our runs by their index order in the original `runs` vec
+        let mut remaining_updates_by_run = vec![0; runs.len()];
+        let mut runs: Vec<_> = runs
+            .iter()
+            .map(|(batch, parts)| (batch, parts.into_iter()))
+            .collect();
+
+        // populate our heap with the updates from the first part of each run
+        for (index, (batch, parts)) in runs.iter_mut().enumerate() {
+            if let Some(key) = parts.next() {
+                fetch_batch_part(
+                    &shard_id,
+                    blob.as_ref(),
+                    &metrics,
+                    key,
+                    &batch.desc,
+                    |k, v, mut t, d| {
+                        t.advance_by(desc.since().borrow());
+                        let d = D::decode(d);
+                        let k = k.to_vec();
+                        let v = v.to_vec();
+                        // default heap ordering is descending
+                        heap.push(Reverse((((k, v), t, d), index)));
+                        remaining_updates_by_run[index] += 1;
+                    },
+                )
+                .await?;
+            }
+        }
+
+        // repeatedly pull off the least element from our heap, refilling from the originating run
+        // if needed. the heap will be exhausted only when all parts from all input runs have been
+        // consumed.
+        while let Some(Reverse((((k, v), t, d), index))) = heap.pop() {
+            remaining_updates_by_run[index] -= 1;
+            // if we've pulled off all the updates from a particular part,
+            // fetch the next part from its originating run
+            if remaining_updates_by_run[index] == 0 {
+                let (batch, parts) = &mut runs[index];
+                if let Some(key) = parts.next() {
+                    fetch_batch_part(
+                        &shard_id,
+                        blob.as_ref(),
+                        &metrics,
+                        key,
+                        &batch.desc,
+                        |k, v, mut t, d| {
+                            t.advance_by(desc.since().borrow());
+                            let d = D::decode(d);
+                            let k = k.to_vec();
+                            let v = v.to_vec();
+                            heap.push(Reverse((((k, v), t, d), index)));
+                            remaining_updates_by_run[index] += 1;
+                        },
+                    )
+                    .await?;
+                }
+            }
+
+            // both T and D are Codec64, ergo 8 bytes a piece
+            update_buffer_size_bytes += k.len() + v.len() + 16;
+            update_buffer.push(((k, v), t, d));
+
+            // if we're more than 90% into our target size, write out our part. this is an
```

Comment:
oooooof. thoughts on instead setting the value we give to the ColumnarRecordsVecBuilder to something large with a comment that we defer the chunking to the compaction code?

Reply:
oh I like that. will update / think about that one a while to see if it all fits together

Reply:
done